src/FixedGuns.cpp

```diff
@@ -255,21 +254,66 @@ void FixedGuns::UpdateLead(float timeStep, int num, Body *ship, Body *target)
 
 	// don't calculate lead if there's no gun there
 	if (m_gun_present[num] && projspeed > 0) {
-		const vector3d targvel = target->GetVelocityRelTo(ship) * ship->GetOrient();
-		vector3d leadpos = targpos + targvel * (targpos.Length() / projspeed);
-		leadpos = targpos + targvel * (leadpos.Length() / projspeed); // second order approx
-
-		m_targetLeadPos = leadpos;
+		if (m_gun[num].projData.beam) {
+			//For beems just shoot where the target is - no lead needed
+			m_targetLeadPos = targpos;
+		} else {
+			const vector3d targvel = target->GetVelocityRelTo(ship) * ship->GetOrient();
+
+#if 0
+			vector3d leadpos = targpos + targvel * (targpos.Length() / projspeed);
+			leadpos = targpos + targvel * (leadpos.Length() / projspeed); // second order approx
+
+			m_targetLeadPos = leadpos;
+#else
+			//Exact lead calculation. We start with:
+			// |targpos * l + targvel| = projspeed
+			//we solve for l which can be interpreted as 1/time for the projectile to reach the target
+			//it gives:
+			// |targpos|^2 * l^2 + targpos*targvel * 2l + |targvel|^2 - projspeed^2 = 0;
+			// so it gives scalar quadratic equation with two posible solutions - we care only about the positive one - shooting forward
+			// A basic math for solving, there is probably more elegant and efficient way to do this:
+			double a = targpos.LengthSqr();
+			double b = targpos.Dot(targvel) * 2;
+			double c = targvel.LengthSqr() - projspeed * projspeed;
+			double delta = b * b - 4 * a * c;
+
+			if (delta < 0) {
+				//no solution
+				m_currentLeadDir = forwardVector;
+				return;
+			}
+
+			double l = (-b + sqrt(delta)) / 2 / a;
+
+			if (l < 0) {
+				//no positive solution
+				m_currentLeadDir = forwardVector;
+				return;
+			} else {
+				//t = 1/l
+				//This is an exact solution as opposed to 2 step aproximation used before
+				//It does not improve the accuracy as expected though.
+				//If the target is accelerating and is far enough then this aim assist will
+				//actually make sure that it is mpossible to hit..
```

Comment:
I'd be interested in a bit more data on the perceptual or actual measured increase in accuracy between the two methods, as the comment here seems to imply both have about the same effective accuracy when considered without the acceleration term.

Reply:
Well I tried to solve this without approximation, lets say, for fun. Theoretically it should be better but in practice there was no noticeable difference. This is why I left the approximation method to decide during review which to choose. I do not think that the exact method is heavier computationalylly so maybe it could stay.

The adjustment for acceleration made noticeable difference. If a target flies away from you and is accelerating then you will get reliable hits. The good thing about the exact method is that it gave me the time to use for the acc adjustment.

Reply:
So the approximation method used sqrt also to calculate length of the distance vector twice, but I suppose if the compiler is not too dumb it would only be done once.
I'll try to get rid of those divisions then. Good to know about the difference in complexities.

Reply:
Hah, good eyes. I can only apologize for being very tired when I wrote that, this is indeed one `sqrt()` cheaper than the prior solution, and the current iteration of the code has the same number of divisions. I'm satisfied with the current state of the code; I don't think there's any further optimization gain here that would be worth the time investment to realize (if at all).

The overall performance cost of this code is quite small for a single iteration (I'd guess around 1-2 microseconds at most), but in practice I'd like to be able to run lead computations on AI turrets, fixed/gimballed ship weapons, etc. for at least 100 (if not closer to 1000) individual weapons per frame in the future, which requires the individual iteration cost to be quite small indeed. :smile:

Reply:
> AI turrets, fixed/gimballed ship weapons, etc. for at least 100 (if not closer to 1000) individual weapons per frame in the future

Wow. Are we gonna have battleships in the game?

Reply:
Eventually stations will need self-defense weapons, and there has long been a plan for large (2km+) "bulk ships"... including military/naval variants of such. You can imagine how many guns would by design be present on such installations!